Basic Wumborian Fugu & Fugu Gland

Jacquerel opened the pull request:
## About The Pull Request

Fixes #72677 and also converted the "Wumborian Fugu" mob to a basic mob rather than a simple one.
I will be totally honest: I didn't need to do that in order to fix the bug. I just didn't like looking at the rest of the code in that file.
Also I have some kind of sickness which makes me do this.

This ended up being one of those "see something related and fix it as well" ones so there's a couple of only tangentially related changes in here. If you want me to split it up I will but I think this one is _probably_ fine because the wide-ranging changes are pretty simple ones?

So what this PR does is:
- Refactors simple mob into basic mob.
- Cleans up its really ugly ability to work in a hopefully nicer way.
- A one line fix to the linked issue above.
- Modifies the default cooldown on `basic_melee_attack` and `attack_obstructions` to be a widely used cooldown rather than a random value used by no mob that we have.
- Renamed behaviour "try_mob_ability" to "targeted_mob_ability" and added a new AI behaviour called "use_mob_ability", the difference between the two being that the former requires a target and the latter does not. I... don't actually use this because I realised after adding it that I still want a target for this mob, but someone will need it eventually.
- Change everywhere that is passing references to abilities to actions to pass weak references instead.
- Adds an element to handle "spawn this stuff when a related mob dies".
- Found a few places where people were setting `environment_smash ` as if it did anything (including me) and replaced them with the proper ai_controller implementation instead, updated the comment to make it clearer although that won't prevent copy/paste errors.
- Registered to the "movement speed updated" signal to ensure that basic mobs actually notice that you have applied a movement speed modifier.

## Why It's Good For The Game

Fixes a linked issue.
Refactors some code which made me sad whenever I saw it.
Restores some mob behaviour which nobody noticed was missing, but was.
Fixes some apparently unreliable code I added in a recent PR reliant on basic mobs using movespeed modifiers.
Adds element we will definitely need again in the future.

## Changelog

:cl:

fix: The Fugu Gland can once more be used on Ian, Carp, Giant Spiders, or other basic mobs.
fix: Syndicate mobs will once again attack windows to try to reach you, and space ruin spiders won't.
fix: Netherworld-themed mobs will correctly adjust their speed as they take damage.
refactor: Made the Wumborian Fugu into a basic mob, which should act largely the same way but may have slightly different speed and reaction times.
/:cl:

Jacquerel committed:
Starts converting

Jacquerel committed:
Merge branch 'master' into more-fish-stuff

Jacquerel committed:
Set default attack/attack obstruction cooldown to the actual most common value
Remove unused "environment_smash" variable and make things which seemed to expect it smash obstacles

Jacquerel committed:
Fixes unclosed bracket

Jacquerel committed:
Breaks up the file
Adds the AI
Rename 'try mob ability' to 'targeted mob ability' and adds 'use mob ability'
Pass abilities as weakrefs instead of hard refs
Adds loot element
Sadly puts environment_smash back in
Basic Mobs will listen to speed modifiers

Jacquerel committed:
UpdatePaths

Jacquerel committed:
Updates comment more

Jacquerel committed:
Update UpdatePath Path

san7890 commented on code/datums/elements/loot_drop.dm:
```diff
@@ -0,0 +1,33 @@
+v/**
```
```suggestion
/**
```

san7890 commented on code/modules/mob/living/basic/space_fauna/wumborian_fugu/fugu_gland.dm:
```diff
@@ -0,0 +1,42 @@
+		balloon_alert(user, "it's dead!")
+		return
+	if(HAS_TRAIT(animal, TRAIT_FUGU_GLANDED))
+		balloon_alert(user, "already wumboed!")
```
```suggestion
		balloon_alert(user, "already large!")
```

it's funny but it doesn't feel intuitive

san7890 commented on code/modules/mob/living/basic/space_fauna/wumborian_fugu/inflation.dm:
```diff
@@ -0,0 +1,92 @@
+		return FALSE
+	if(HAS_TRAIT(owner, TRAIT_FUGU_GLANDED))
+		if (feedback)
+			owner.balloon_alert(owner, "ALREADY WUMBO!")
```
```suggestion
			owner.balloon_alert(owner, "already large!")
```

i think we don't want capitialization in balloon alerts anyways

san7890 commented on code/modules/mob/living/basic/space_fauna/wumborian_fugu/wumborian_fugu.dm:
```diff
@@ -0,0 +1,57 @@
+	minimum_survivable_temperature = 0
+	maximum_survivable_temperature = INFINITY
+	ai_controller = /datum/ai_controller/basic_controller/wumborian_fugu
+	var/datum/action/cooldown/fugu_expand/expand
```
DMDoc with something simple otherwise our stats are hurt

san7890 commented on tools/UpdatePaths/Scripts/73415_simple_fugu_to_basic.txt:
```diff
@@ -0,0 +1,3 @@
+# replace simple animal fugu with basic animal
+
+/mob/living/simple_animal/hostile/asteroid/fugu/@SUBTYPES : /mob/living/basic/wumborian_fugu/@SUBTYPES{@OLD}
```
it's probably fine to leave it as subtypes but in the future, updatepaths should be written for expected changes (unless you know of a place that actually subtyped this fuckin' mob)

san7890 commented on code/modules/mob/living/basic/space_fauna/wumborian_fugu/inflation.dm:
```diff
@@ -0,0 +1,92 @@
+	owner.melee_damage_lower = 15
+	owner.melee_damage_upper = 20
+	owner.status_flags |= GODMODE
+	var/mob/living/basic/basic_owner = owner
+	basic_owner.obj_damage = 60
+	basic_owner.ai_controller.blackboard[BB_BASIC_MOB_FLEEING] = FALSE
+	basic_owner.ai_controller.CancelActions()
```
Aren't there cases in which we could theoretically see this effect on a non-basic mob (i.e. someone in the future adds this as their cool new action)? If so, add a typecheck just in case we don't modify non-existent vars on a carbon or something

Jacquerel replied:
I tried to figure out a way to enable those future cases but in this case the behaviour is _too_ specific.
There's already a type check, the ability simply fails to trigger if you're not the appropriate kind of mob when you use it.
Additionally, the status effect fails to apply to anything which isn't the correct type of mob as well.

I got as far as:
```
owner.add_movespeed_modifier(/datum/movespeed_modifier/status_effect/inflated)
ADD_TRAIT(owner, TRAIT_FUGU_GLANDED, TRAIT_STATUS_EFFECT(id))
owner.AddElement(/datum/element/wall_smasher)
owner.mob_size = MOB_SIZE_LARGE
```
but anything under that was just too specific to this one use case for me to have a sensible way of preserving the behaviour while also making it a sane option for any other mob.

Jacquerel replied:
This is notably also true of the previous incarnation's ability, but that one for some reason had _half_ of the implementation inside an ability and then half in a callback on the mob? Reading it was what drove me to convert the mob in the first place.

Jacquerel replied:
It was already type guarded for being a `/mob/living/basic` by the line in `on_creation` but this has made it clear to me that the _code_ doesn't describe that so I've rewritten it in a way that if anyone touches this in the future it will be transparent that this only works on one specific kind of mob and if they change that they will have to change other things.

san7890 reviewed (changes requested):
few things on first look

Jacquerel committed:
Removes random trailing character
Desouls balloon alerts
Adds missing documenting comment

Jacquerel committed:
Updates UpdatePath Content

Jacquerel committed:
Lint file order
Update type guards so that if someone refactors this in the future they can't accidentally make it runtime

Mothblocks commented on code/datums/elements/loot_drop.dm:
```diff
@@ -0,0 +1,33 @@
+	for (var/typepath in loot)
+		new typepath(source.loc)
+	if (only_once)
+		source.RemoveElement(/datum/element/loot_drop)
```
Don't you need to include the arguments here because it is bespoke?

Jacquerel committed:
pass arguments to remove element

san7890 commented on code/datums/elements/loot_drop.dm:
```diff
@@ -0,0 +1,33 @@
+ *
+ * A mob with this element will create the provided atoms when it dies.
+ */
+/datum/element/loot_drop
```
I was wondering why you included a new element because I swore I knew of bileworms dropping loot, and then it hit me: Why don't we just extend `/datum/element/death_drops`? Seems kinda silly to have two elements that do the exact same thing...

Jacquerel committed:
Removes unecessary extra element
